packages/grid/x-data-grid-pro/src/components/DataGridProVirtualScroller.tsx

```diff
@@ -170,8 +170,8 @@ const VirtualScrollerPinnedRows = styled('div', {
   const boxShadowColor = getBoxShadowColor(theme);
   return {
     position: 'sticky',
-    // should be above the detail panel
-    zIndex: 3,
+    // should be above the no rows overlay
+    zIndex: 5,
```

Comment:
Should we increase zIndex for loading overlay?

<img width="808" src="https://github.com/mui/mui-x/assets/13808724/9f9da001-e49e-46d6-9e2d-a593946e5462">

Reply:
It's not going to work that way. Both the loading and the no rows overlays use the same container that has the `z-index`

Reply:
We can change the zIndex depending on the overlay type, something like this:

```diff
diff --git a/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx b/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
index 1b1068045..7714e7c14 100644
--- a/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
+++ b/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
@@ -21,14 +21,17 @@ const GridOverlayWrapperRoot = styled('div', {
   name: 'MuiDataGrid',
   slot: 'OverlayWrapper',
   overridesResolver: (props, styles) => styles.overlayWrapper,
-})({
+})<{ overlayType: 'loadingOverlay' | string }>(({ overlayType }) => ({
   position: 'sticky', // To stay in place while scrolling
   top: 0,
   left: 0,
   width: 0, // To stay above the content instead of shifting it down
   height: 0, // To stay above the content instead of shifting it down
-  zIndex: 4, // Should be above pinned columns, pinned rows and detail panel
-});
+  zIndex:
+    overlayType === 'loadingOverlay'
+      ? 5 // Should be above pinned columns, pinned rows, and detail panel
+      : 4, // Should be above pinned columns and detail panel
+}));
 
 const GridOverlayWrapperInner = styled('div', {
   name: 'MuiDataGrid',
@@ -49,7 +52,7 @@ const useUtilityClasses = (ownerState: OwnerState) => {
   return composeClasses(slots, getDataGridUtilityClass, classes);
 };
 
-function GridOverlayWrapper(props: React.PropsWithChildren<{}>) {
+function GridOverlayWrapper(props: React.PropsWithChildren<{ overlayType: string }>) {
   const apiRef = useGridApiContext();
   const rootProps = useGridRootProps();
 
@@ -77,7 +80,7 @@ function GridOverlayWrapper(props: React.PropsWithChildren<{}>) {
   }
 
   return (
-    <GridOverlayWrapperRoot className={clsx(classes.root)}>
+    <GridOverlayWrapperRoot className={clsx(classes.root)} overlayType={overlayType}>
       <GridOverlayWrapperInner
         className={clsx(classes.inner)}
         style={{
@@ -102,16 +105,20 @@ export function GridOverlays() {
   const showNoResultsOverlay = !loading && totalRowCount > 0 && visibleRowCount === 0;
 
   let overlay: React.JSX.Element | null = null;
+  let overlayType = '';
 
   if (showNoRowsOverlay) {
+    overlayType = 'noRowsOverlay';
     overlay = <rootProps.slots.noRowsOverlay {...rootProps.slotProps?.noRowsOverlay} />;
   }
 
   if (showNoResultsOverlay) {
+    overlayType = 'noResultsOverlay';
     overlay = <rootProps.slots.noResultsOverlay {...rootProps.slotProps?.noResultsOverlay} />;
   }
 
   if (loading) {
+    overlayType = 'loadingOverlay';
     overlay = <rootProps.slots.loadingOverlay {...rootProps.slotProps?.loadingOverlay} />;
   }
 
@@ -119,5 +126,5 @@ export function GridOverlays() {
     return null;
   }
 
-  return <GridOverlayWrapper>{overlay}</GridOverlayWrapper>;
+  return <GridOverlayWrapper overlayType={overlayType}>{overlay}</GridOverlayWrapper>;
 }

```

Reply:
Got it. Updated

Reply:
Personal preference but I don't like too much using strings as constants. I generally do something like this:

```typescript
enum OverlayType {
  NoRows,
  NoResults,
  Loading,
}
```

Reply:
Yeah, I feel you. We had this argument so many times that idk to be honest which one is the better way. I honestly would opt for no strings but I know we had an unwritten rule that if something is used only once then we use strings (at least this was in the past).

Reply:
I prefer enums over unions because enums (without default values) use integers instead of strings. In many cases (this one included) it doesn't make much of a difference, but string comparisons can bring the performance down if used carelessly. [Here](https://github.com/json5/json5/pull/278) is an admittedly very specific use-case where string usage causes a 2x slowdown.

I also find enums easier to reason about, they feel less like a hack imho.

Reply:
> https://github.com/json5/json5/pull/278 is an admittedly very specific use-case where string usage causes a 2x slowdown.

Nice, great work! I'm surprised it hasn't been merged yet 😅

> I also find enums easier to reason about, they feel less like a hack IMHO.

I'm fine with using both internally depending on the use case and personal preference 👍🏻

When it comes to the public API and DX, how do you feel about this?

```tsx
import { DataGrid } from '@mui/x-data-grid';
// string union
<DataGrid filterMode="server" />
```

```tsx
import { DataGrid, GridFilterMode } from '@mui/x-data-grid';
// enum
<DataGrid filterMode={GridFilterMode.Server} />
```

I definitely prefer string union, because:
1. No need to import any additional types
2. Autocomplete

Reply:
@MBilalShafi feel free to chime in 🙂

Reply:
Some really good arguments made and awesome work with the `parse()` performance PR Rom 🚀 

I kind of agree with preferring the unions, at least for the public API as the users will not have to learn what's a specific enum is called and import it, filterMode="server" is a good example.

But when there's already a learning curve, like with the overlay names, which we define and if customization is needed users will have to know the names anyways. In such cases, enums could make more sense as it looks more readable on top of the slight performance benefit it gives by using integers instead of strings.

Reply:
> Nice, great work! I'm surprised it hasn't been merged yet 😅

I had a feeling the maintainer wouldn't merge. It's good for performance, bad for readability. Did it mostly for fun though :sweat_smile: 

For the public API, I'll agree that unions are more ergonomic. JSX props feel more natural with a string than with an enum.

Internally, for the large majority of cases it won't matter, but imho it would be a good habit to use enums because in the few cases where it matters, we'll already be using the efficient solution.